src/colvargrid_integrate.cpp

```diff
@@ -146,92 +189,187 @@ void colvargrid_integrate::update_div_neighbors(const std::vector<int> &ix0)
-      / widths[2]) * 0.25;
+        ((g100[0] - g000[0] + g101[0] - g001[0] + g110[0] - g010[0] + g111[0] - g011[0]) /
+             widths[0] +
+         (g010[1] - g000[1] + g011[1] - g001[1] + g110[1] - g100[1] + g111[1] - g101[1]) /
+             widths[1] +
+         (g001[2] - g000[2] + g011[2] - g010[2] + g101[2] - g100[2] + g111[2] - g110[2]) /
+             widths[2]) *
+        0.25;
+    cvm::real supplement = 0;
+    cvm::real supplement_coefficient = 1;
+
+    ix = ix0;
+    if ((ix0[0] == 0 || ix0[0] == computation_nx[0] - 1) && !periodic[0]) {
+      cvm::real coefficient = (ix[0] == 0) ? 1.0 : -1.0;
+      supplement_coefficient *= 0.5;
+      supplement = coefficient *
+                   (g100[0] + g000[0] + g101[0] + g001[0] + g110[0] + g010[0] + g111[0] + g011[0]) *
+                   0.125 * 2 * widths[0];
+    }
+    if ((ix0[1] == 0 || ix0[1] == computation_nx[1] - 1) && !periodic[1]) {
+      cvm::real coefficient = (ix[1] == 0) ? 1.0 : -1.0;
+      supplement_coefficient *= 0.5;
+      supplement +=
+          coefficient *
+          (g010[1] + g000[1] + g011[1] + g001[1] + g110[1] + g100[1] + g111[1] + g101[1]) *
+          widths[1] * 2 * 0.125;
+    }
+    if ((ix0[2] == 0 || ix0[2] == computation_nx[2] - 1) && !periodic[2]) {
+      cvm::real coefficient = (ix[2] == 0) ? 1.0 : -1.0;
+      supplement_coefficient *= 0.5;
+      supplement +=
+          coefficient *
+          (g001[2] - g000[2] + g011[2] - g010[2] + g101[2] - g100[2] + g111[2] - g110[2]) * 0.125 *
+          2 * widths[2];
+    }
+    divergence[linear_index] += supplement;
+    divergence[linear_index] *= supplement_coefficient;
   }
 }
 
 
+void colvargrid_integrate::prepare_divergence_stencils()
+{
+  surrounding_points_relative_positions.clear();
+  size_t n_combinations = pow(2, nd);
+  for (size_t i = 0; i < n_combinations; i++) {
+    surrounding_points_relative_positions.push_back(convert_base_two(i, nd));
+  }
+}
+
+/// Updates the divergence at the point ix0
+void colvargrid_integrate::update_weighted_div_local(const std::vector<int> &ix0)
+{
+  const size_t linear_index = computation_grid->address(ix0);
+  cvm::real div_at_point = 0;
+  for (std::vector<int> surrounding_point_relative_position :
+       surrounding_points_relative_positions) {
+    std::vector<int> surrounding_point_coordinates = ix0;
+    std::vector<cvm::real> gradient_at_surrounding_point(nd, 0);
+    for (size_t i = 0; i < nd; i++) {
+      surrounding_point_coordinates[i] += surrounding_point_relative_position[i];
+    }
+
+    gradients->wrap_detect_edge(surrounding_point_coordinates);
+    get_regularized_grad(gradient_at_surrounding_point, surrounding_point_coordinates);
+    cvm::real weight = regularized_weights[gradients->address(surrounding_point_coordinates)];
+
+    for (size_t i = 0; i < nd; i++) {
+      div_at_point += pow(-1, surrounding_point_relative_position[i] + 1) *
```

Comment:
(-1)^n could be either -1 or 1 depending on whether n is an even number or an odd number, so this can be simply calculated without calling `pow`.